src/template.rs

```diff
@@ -0,0 +1,142 @@
+use std::fs::File;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+
+use xml::attribute::OwnedAttribute;
+use xml::reader::XmlEvent;
+use xml::EventReader;
+
+use crate::error::TiledError;
+use crate::{
+    util::*, EmbeddedParseResultType, MapTilesetGid, ObjectData, ParseContext, ResourceCache,
+    TemplateOverride, Tileset,
+};
+
+/// A template, consisting of an object and a tileset
+///
+/// The path is the path to the template file; it is used to identify them uniquely, and skip parsing the file multiple
+/// times.
+#[derive(Clone, Debug, PartialEq)]
+pub(crate) struct Template {
+    pub path: PathBuf,
+    pub tileset: Option<Arc<Tileset>>,
+    pub object: Option<ObjectData>,
+}
+
+impl Template {
+    pub(crate) fn new_external<CacheTy: ResourceCache>(
+        path: &Path,
+        parse_context: &mut ParseContext<CacheTy>,
+    ) -> Result<usize, TiledError> {
+        // First, check if this template already exists in our list of templates (it will be uniquely identified by its
+        // path).
+        if let Some((id, _)) = parse_context
+            .templates
+            .iter()
+            .enumerate()
+            .find(|(_, template)| template.path == path)
+        {
+            return Ok(id);
+        }
+
+        // Open the template file
+        let file = File::open(&path).map_err(|err| TiledError::CouldNotOpenFile {
+            path: path.to_path_buf(),
+            err,
+        })?;
+
+        let mut tileset_parser = EventReader::new(file);
+        loop {
+            match tileset_parser
+                .next()
+                .map_err(TiledError::XmlDecodingError)?
+            {
+                XmlEvent::StartElement {
+                    name, attributes, ..
+                } if name.local_name == "template" => {
+                    return Self::parse_external_template(
+                        &mut tileset_parser.into_iter(),
+                        &attributes,
+                        path,
+                        parse_context,
+                    );
+                }
+                XmlEvent::EndDocument => {
+                    return Err(TiledError::PrematureEnd(
+                        "Tileset Document ended before map was parsed".to_string(),
+                    ))
+                }
+                _ => {}
+            }
+        }
+    }
+
+    fn parse_external_template<'a, CacheTy: ResourceCache>(
+        parser: &mut impl Iterator<Item = XmlEventResult>,
+        _attrs: &Vec<OwnedAttribute>,
+        path: &Path,
+        parse_context: &mut ParseContext<CacheTy>,
+    ) -> Result<usize, TiledError> {
+        let mut object = Option::None;
+
+        let template_id = parse_context.templates.len();
+        parse_context.templates.push(Template {
+            path: path.to_path_buf(),
+            tileset: None,
+            object: None,
+        });
```

Comment:
Again, this shouldn't be here; These are changes you wouldn't expect from a function called `parse_external_template`.